wgsl/index.bs

```diff
@@ -395,21 +411,103 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
 * [[#array-types]] if the array type has a stride attribute and its element type is host-shareable
 * [[#struct-types]] if all its members are host-shareable
 
+The contents of uniform buffers and storage buffers must be of host-shareable type.
+
 TODO: *This is a stub*: Collectively, the *stride* and *offset* attributes are called *layout attributes*.
 
 ### Storage Classes TODO ### {#storage-class}
 
+Memory locations are partitioned into <dfn noexport>storage classes</dfn>.
+Each storage class has unique properties affecting mutability, visibility, and the values they may contain.
+
+
+<table class='data'>
+  <thead>
+    <tr><td>Storage class
+        <td>Readable by shader?<br>Writable by shader?
+        <td>Sharing among invocations
+        <td>Scope
+        <td>Restrictions on stored values
+        <td>Notes
+  </thead>
+  <tr><td>input
+      <td>Read-only
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=IO-shareable=]
+      <td>Contains values written by an upstream pipeline stage, or by the implementation.
+  <tr><td>output
+      <td>Read-write
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=IO-shareable=]
+      <td>Shader writes values for downstream pipeline processing.
+  <tr><td>function
+      <td>Read-write
+      <td>Same invocation only
+      <td>[=Function scope=]
+      <td>[=Storable=]
+      <td>
+  <tr><td>private
+      <td>Read-write
+      <td>Same invocation only
+      <td>[=Module scope=]
+      <td>[=Storable=]
+      <td>
+  <tr><td>workgroup
+      <td>Read-write
+      <td>Invocations in the same [=workgroup=]
+      <td>[=Module scope=]
+      <td>[=Storable=]
+      <td>
+  <tr><td>uniform
+      <td>Read-only
+      <td>Invocations in the same [=shader stage=]
+      <td>[=Module scope=]
+      <td>[=Host-shareable=]
+      <td>See [[#memory-layout]]
+  <tr><td>storage_buffer
+      <td>Readable.<br>
+          Also writable if the variable is not read-only.
+      <td>Invocations in the same [=shader stage=]
+      <td>[=Module scope=]
+      <td>[=Host-shareable=]
+      <td>See [[#memory-layout]].
```

Comment:
I find it confusing that we have 3 different uniform-ish storage classes. Was hoping to navigate to that `[[#memory-layout]` to understand more, but it's empty :)

Reply:
Right, the memory-layout section is empty right now. That is yet another big change I would like to do separately.

There are 3 uniform-ish things:
- storage_buffer.  For OpenGL's SSBOs. Read-write buffers with explicit layout. Contents shareable with the host. Pretty flexible layout rules 
      - What's confusing is that originally Vulkan/SPIR-V had this in "Uniform" storage class with "BufferBlock" decorated struct.  But now the style is "StorageBuffer" with Block-decorated struct
- uniform. For OpenGL's UBOs.  Read-only buffer with explicit layout. Strict and annoying array layouts. Limit max size. Contents shareable with host.
- uniform_constant.  For opaque things that happen to be handles, each with their own binding. The handles are immutable, but the underlying data might not be.    "UniformConstant" is not my favourite name in the world but I suggest we bikeshed that separately.

Reply:
This separation is rather confusing. First of all, it doesn't include storage images. I [tried to check](http://shader-playground.timjones.io/427de82dab2d4ae80fba7a792f018a6d) which storage class does glslang assign to them, and apparently I'm getting "UniformConstant" for a write-only storage image... It makes so little sense to me.

I guess the bigger question to ask is - why does WGSL need storage classes *at all*?

Reply:
For one thing, you need to separate workgroup storage from private, for one thing.

Also, you have different rules for different storage classes. Layout rules differ.  

Fundamentally, storage classes are a way of organizing the many rules about how memory treatedl